pkg/reconciler/cache/replication/replication_reconcile_unstructured.go

```diff
@@ -0,0 +1,213 @@
+	}
+	localObjMetaRaw, hasLocalObjMetaRaw, err := unstructured.NestedFieldNoCopy(unstructuredLocalObject.Object, "metadata")
+	if err != nil {
+		return false, err
+	}
+	localObjMeta, ok := localObjMetaRaw.(map[string]interface{})
+	if !ok {
+		return false, fmt.Errorf("metadata field of unstructuredLocalObjectMeta is of the type %T, expected map[string]interface{}", localObjMetaRaw)
+	}
+	if !hasLocalObjMetaRaw && !hasCacheObjMetaRaw {
+		return false, nil // no-op
+	}
+	if !hasLocalObjMetaRaw {
+		unstructured.RemoveNestedField(unstructuredCacheObject.Object, "metadata")
+		return true, nil
+	}
+
+	// before we can compare the cache object we need to
+	// store, remove and then bring back fields that are unique only to the cache object
+	if cacheObjRV, hasCacheObjRV := cacheObjMeta["resourceVersion"]; hasCacheObjRV {
+		unstructured.RemoveNestedField(cacheObjMeta, "resourceVersion")
+		defer func() {
+			if err == nil {
+				err = unstructured.SetNestedField(unstructuredCacheObject.Object, cacheObjRV, "metadata", "resourceVersion")
+			}
+		}()
+	}
+	if cacheObjAnnotationsRaw, hasCacheObjAnnotations := cacheObjMeta["annotations"]; hasCacheObjAnnotations {
+		cacheObjAnnotations, ok := cacheObjAnnotationsRaw.(map[string]interface{})
+		if !ok {
+			return false, fmt.Errorf("metadata.annotations field of unstructuredCacheObject is of the type %T, expected map[string]interface{}", cacheObjAnnotationsRaw)
+		}
+		if shard, hasShard := cacheObjAnnotations[genericrequest.AnnotationKey]; hasShard {
+			unstructured.RemoveNestedField(cacheObjAnnotations, genericrequest.AnnotationKey)
+			defer func() {
+				if err == nil {
+					err = unstructured.SetNestedField(unstructuredCacheObject.Object, shard, "metadata", "annotations", genericrequest.AnnotationKey)
+				}
+			}()
+		}
+		// TODO: in the future the original RV will be stored in an annotation
+	}
+
+	// before we can compare with the local object we need to
+	// store, remove and then bring back the ResourceVersion on the local object
+	if localObjRV, hasLocalObjRV := localObjMeta["resourceVersion"]; hasLocalObjRV {
+		unstructured.RemoveNestedField(localObjMeta, "resourceVersion")
+		defer func() {
+			if err == nil {
+				localObjMeta["resourceVersion"] = localObjRV
+			}
+		}()
+	}
+
+	metaChanged = !reflect.DeepEqual(cacheObjMeta, localObjMeta)
+	if !metaChanged {
+		return false, nil
+	}
+
+	newCacheObjMeta := map[string]interface{}{}
+	for k, v := range localObjMeta {
+		newCacheObjMeta[k] = v
+	}
+	return true, unstructured.SetNestedMap(unstructuredCacheObject.Object, newCacheObjMeta, "metadata")
+}
+
+// ensureUnstructuredRemaining changes unstructuredCacheObject to match unstructuredLocalObject except for the metadata field
+// returns true when the unstructuredCacheObject was updated.
+func ensureUnstructuredRemaining(unstructuredCacheObject *unstructured.Unstructured, unstructuredLocalObject *unstructured.Unstructured) (bool, error) {
```

Comment:
everything in here is unstructured. Drop the prefix in the local vars.

change the long "found"-vars into just `found`, e.g. `hasUnstructuredLocalObjectMeta -> found`.